targets/TARGET_STM/stm_spi_api.c

```diff
@@ -777,75 +800,225 @@ static inline int datasize_to_transfer_bitshift(uint32_t DataSize)
+        /* Set transaction size and run SPI */
+        LL_SPI_SetTransferSize(SPI_INST(obj), rx_length);
+        LL_SPI_Enable(SPI_INST(obj));
+        LL_SPI_StartMasterTransfer(SPI_INST(obj));
+
+        /* Receive data */
+        for (int i = 0; i < rx_length; i++) {
+            msp_wait_readable(obj);
+            rx_buffer[i] = msp_read_data(obj, bitshift);
+        }
+
+        /* Stop SPI */
+        LL_SPI_Disable(SPI_INST(obj));
+        /* Clear transaction flags */
+        LL_SPI_ClearFlag_EOT(SPI_INST(obj));
+        LL_SPI_ClearFlag_TXTF(SPI_INST(obj));
+        /* Reset transaction size */
+        LL_SPI_SetTransferSize(SPI_INST(obj), 0);
+#else  /* TARGET_STM32H7 */
+        /* Unlike STM32H7 other STM32 families generates SPI Clock signal continuously in half-duplex receive mode
+         * till SPI is enabled. To stop clock generation a SPI should be disabled during last frame receiving,
+         * after generation at least one SPI clock cycle. It causes necessity of critical section usage.
+         * So the following consequences of steps is used to receive each byte:
+         * 1. Enter into critical section.
+         * 2. Enable SPI.
+         * 3. Wait one SPI clock cycle.
+         * 4. Disable SPI.
+         * 5. Wait full byte receiving.
+         * 6. Read byte.
+         * It gives some overhead, but gives stable byte reception without dummy reads and
+         * short delay of critical section holding.
+         */
+
+        /* get estimation about one SPI clock cycle */
+        uint32_t baudrate_period_ns = 1000000000 / spi_get_baudrate(obj);
+
+        for (int i = 0; i < rx_length; i++) {
+            core_util_critical_section_enter();
+            LL_SPI_Enable(SPI_INST(obj));
+            /* Wait single SPI clock cycle. */
+            wait_ns(baudrate_period_ns);
+            LL_SPI_Disable(SPI_INST(obj));
+            core_util_critical_section_exit();
+
+            msp_wait_readable(obj);
+            rx_buffer[i] = msp_read_data(obj, bitshift);
+        }
+
+#endif /* TARGET_STM32H7 */
+    }
+
+    return rx_length + tx_length;
+}
+
+int spi_master_write(spi_t *obj, int value)
+{
+    struct spi_s *spiobj = SPI_S(obj);
+    SPI_HandleTypeDef *handle = &(spiobj->handle);
+
+    if (handle->Init.Direction == SPI_DIRECTION_1LINE) {
+        int result = spi_master_one_wire_write(obj, (char *)&value, 1, NULL, 0);
+        return result == 1 ? 0 : 1;
+    }
+    const int bitshift = datasize_to_transfer_bitshift(handle->Init.DataSize);
+    MBED_ASSERT(bitshift >= 0);
+
+    /*  Here we're using LL which means direct registers access
+     *  There is no error management, so we may end up looping
+     *  infinitely here in case of faulty device for instance,
+     *  but this will increase performances significantly
+     */
+
+#if TARGET_STM32H7
+    /* Master transfer start */
+    LL_SPI_StartMasterTransfer(SPI_INST(obj));
+#endif /* TARGET_STM32H7 */
+
+    /* Transmit data */
```

Comment:
I would have preferred that only the 3-wires case is modified.
Or maybe create a first commit that introduce new inline functions, then introduce the 3-wires changes ... ?

Reply:
Done. The changes are splitted into multiple commits (inline functions, 3-wires changes for synchronous API, some improvements for 3-wire with asynchronous API)

Reply:
thanks